File: backend/app/rag/retriever.py
"""
Retriever — orchestrates search with deduplication and context window management.
"""
import logging
from app.rag.vector_store import vector_store
from app.config import get_settings

logger = logging.getLogger(__name__)
settings = get_settings()


class Retriever:

    def __init__(self):
        self.vector_store = vector_store
        self.min_score_threshold = 0.3

    async def retrieve(self, query: str, user_id: str, document_ids: list[str] = None, top_k: int = None) -> list[dict]:
        top_k = top_k or settings.top_k_retrieval

        raw_results = await self.vector_store.search(
            query=query, user_id=user_id, document_ids=document_ids, top_k=top_k * 2,
        )

        if not raw_results:
            return []

        filtered = [r for r in raw_results if r["score"] >= self.min_score_threshold]
        deduplicated = self._deduplicate(filtered)
        trimmed = self._trim_to_context_window(deduplicated[:top_k])

        logger.info(f"Retrieved {len(trimmed)} chunks (raw={len(raw_results)}, filtered={len(filtered)})")
        return trimmed

    def _deduplicate(self, results: list[dict], similarity_threshold: float = 0.85) -> list[dict]:
        unique_results = []
        seen_texts = []

        for result in results:
            text = result["text"]
            is_duplicate = False

            for seen_text in seen_texts:
                words_a = set(text.lower().split())
                words_b = set(seen_text.lower().split())

                if not words_a or not words_b:
                    continue

                overlap = len(words_a & words_b) / min(len(words_a), len(words_b))
                if overlap > similarity_threshold:
                    is_duplicate = True
                    break

            if not is_duplicate:
                unique_results.append(result)
                seen_texts.append(text)

        return unique_results

    def _trim_to_context_window(self, results: list[dict]) -> list[dict]:
        max_chars = settings.max_context_tokens * 4
        current_chars = 0
        trimmed = []

        for result in results:
            text_len = len(result["text"])
            if current_chars + text_len > max_chars:
                remaining = max_chars - current_chars
                if remaining > 200:
                    result = result.copy()
                    result["text"] = result["text"][:remaining] + "..."
                    trimmed.append(result)
                break
            trimmed.append(result)
            current_chars += text_len

        return trimmed


retriever = Retriever()

File: backend/src/rag/retriever.js
'use strict';
const vectorStore = require('./vectorStore');
const config = require('../config');

const MIN_SCORE = 0.3;

function deduplicate(results, threshold = 0.85) {
  const unique = [];
  const seen = [];

  for (const result of results) {
    const wordsA = new Set(result.text.toLowerCase().split(/\s+/));
    let isDupe = false;

    for (const seenText of seen) {
      const wordsB = new Set(seenText.toLowerCase().split(/\s+/));
      const intersection = [...wordsA].filter(w => wordsB.has(w)).length;
      const overlap = intersection / Math.min(wordsA.size, wordsB.size);
      if (overlap > threshold) { isDupe = true; break; }
    }

    if (!isDupe) {
      unique.push(result);
      seen.push(result.text);
    }
  }
  return unique;
}

function trimToContextWindow(results) {
  const maxChars = config.maxContextTokens * 4;
  let usedChars = 0;
  const trimmed = [];

  for (let result of results) {
    const len = result.text.length;
    if (usedChars + len > maxChars) {
      const remaining = maxChars - usedChars;
      if (remaining > 200) {
        result = { ...result, text: result.text.slice(0, remaining) + '...' };
        trimmed.push(result);
      }
      break;
    }
    trimmed.push(result);
    usedChars += len;
  }
  return trimmed;
}

async function retrieve(query, userId, documentIds = [], topK = null) {
  topK = topK || config.topKRetrieval;

  const raw = await vectorStore.search(query, userId, documentIds, topK * 2);
  if (!raw.length) return [];

  const filtered = raw.filter(r => r.score >= MIN_SCORE);
  const deduped = deduplicate(filtered);
  const trimmed = trimToContextWindow(deduped.slice(0, topK));

  console.log(`[Retriever] ${trimmed.length} chunks (raw=${raw.length}, filtered=${filtered.length})`);
  return trimmed;
}

module.exports = { retrieve };
